Add Dispute Challenge and Accept buttons to Transaction Details

Jinksi opened the pull request:
Fixes #6926 

> **Note**
> This PR is behind feature flag `_wcpay_feature_dispute_on_transaction_page`

#### Changes proposed in this Pull Request

* Adds dispute action buttons to the Transaction Details screen for disputes awaiting a response **(not inquiries)**, enabling the merchant to _Accept_ the dispute or begin/continue a _Challenge_ from this screen.
* Adds a test suite for the `DisputeDetails` component
* Introduces code quality improvements – updated and annotated some TS interfaces.

**Note**: this PR won't handle CTA buttons for inquiries – this will be handled in issue #7193.

**Screenshots**

Action buttons:

![dispute-actions](https://github.com/Automattic/woocommerce-payments/assets/3147296/74cb86af-3df6-4777-90b7-e3a30988c78d)

Accept modal:

![image](https://github.com/Automattic/woocommerce-payments/assets/3147296/bfe01917-4fb7-4653-9654-5e1f822a8031)

Snackbar on successful accept:
<img src="https://github.com/Automattic/woocommerce-payments/assets/3147296/7716d502-36da-482a-aeb9-8aa68f607a92" width="300px" />

Wide buttons on mobile (see design: RiEQmaKRI7u9USAI3lyZz0VX-fi-7430_30973):
<img src="https://github.com/Automattic/woocommerce-payments/assets/3147296/d04e9fb5-9644-4399-a870-941456efd285" width="300px" />

The below video demonstrates:
→ clicking the Challenge button
→ staging evidence via the challenge form
→ returning to the dispute details
→ opening the accept dispute modal
→ accepting the dispute

https://github.com/Automattic/woocommerce-payments/assets/3147296/89f5dc84-8c91-4c80-98b3-1f2604550e55


#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Create a disputed transaction using the card `4000000000000259` at checkout.
* Navigate to the disputed Transaction (Transactions → click the disputed transaction)
* Click **Challenge dispute** → You should be taken to the Dispute challenge screen.
* Click **Accept dispute** →  A modal will be shown. Confirm the text and styles are correct as per the design (RiEQmaKRI7u9USAI3lyZz0VX-fi-6974_70994).
	* Click **Cancel** to close the modal
	* Disable network / go offline in devtools (to force a network error)
	* Click **Accept dispute**
		* On error, a snackbar will show an error message in the bottom-left corner.
	* Re-enable network
	* Click **Accept dispute**
		* The modal should close
		* The dispute should be accepted
		* A snackbar will show a success message in the bottom-left corner
		* The Transaction Details screen will be hydrated with new data (`Disputed: Lost` status chip).
		* The dispute actions should no longer be available.
* View a dispute of any other status (won, lost, under review)
	* The dispute actions should not be visible
* View a disputed transaction for a dispute where a challenge has been staged (click "Save for Later" on the dispute challenge form.
	* The _Challenge_ button should be labelled `Continue with challenge`

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

Jinksi committed:
Refactor isAwaitingResponse check to wrap entire <Card />

Note the final return statement will return the non-awaiting-response UI

Jinksi committed:
Add dispute actions init

Jinksi committed:
Add Dispute > EvidenceDetails type comments to improve devex

Jinksi committed:
Show different challenge button label if has staged evidence

Jinksi committed:
Add missing `payment_intent` to Dispute TS interface

Jinksi committed:
Add action func for accepting disputes on the Transaction Details screen

Jinksi committed:
Add custom hook for accepting dispute

Jinksi committed:
Add accept dispute modal

Jinksi committed:
Improve isLoading state for accepting disputes and exceptions

Jinksi commented on client/types/disputes.d.ts:
```diff
@@ -56,6 +69,7 @@ export interface Dispute {
 	currency: string;
 	created: number;
 	balance_transactions: BalanceTransaction[];
+	payment_intent: string;
```
This was missing from the Dispute TS interface, but is a property of the Stripe API dispute object: https://stripe.com/docs/api/disputes/object#dispute_object-payment_intent.

This PR uses `dispute.payment_intent` to refresh the `payment_intent` data for the Transaction Details when a dispute is accepted.

Jinksi commented on client/data/disputes/actions.js:
```diff
@@ -88,3 +89,58 @@ export function* acceptDispute( id ) {
+// This function handles the dispute acceptance flow from the Transaction Details screen.
+// It will become the default acceptDispute function once the feature flag
+// '_wcpay_feature_dispute_on_transaction_page' is enabled by default.
+export function* acceptTransactionDetailsDispute( dispute ) {
```
Rather than modify the existing `acceptDispute` action (used on the Dispute → Details screen), I've created a new one since there are some differences in the approach:
1. The existing action will force a redirect to Payments → Disputes.
2. This new action requires fetching and updating the payment intent to refresh the current Transaction Details page.

Once the feature flag is lifted, `_wcpay_feature_dispute_on_transaction_page`, we can remove and replace the previous action with the new one.

I've left a TODO item on #6883 to make this change.

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7093` or branch name `fix/6926-transaction-dispute-details-actions` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/6926-transaction-dispute-details-actions) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: eb236055ee50ce3bf83b040efaa5966bdad341dd
- Build time: 2023-09-19 23:24:27 UTC

_Note: the build is updated when a new commit is pushed to this PR._

Jinksi commented on client/types/disputes.d.ts:
```diff
@@ -11,8 +11,21 @@ interface Evidence {
 
 interface EvidenceDetails {
+	/**
+	 * Whether evidence has been staged for this dispute.
```
I wasn't sure what `has_evidence` represented, so I've added explanatory comments to this TS interface to improve code quality / developer experience.

Jinksi committed:
Merge branch 'develop' into fix/6926-transaction-dispute-details-actions

Jinksi committed:
Update accept dispute modal styles

Jinksi committed:
Add getDisputeFee util function

Jinksi committed:
Populate dynamic content for accept dispute modal

Jinksi committed:
Add changelog entry

Jinksi committed:
Fix modal text content overflow

Jinksi committed:
Update Dispute → Charge interface to accept `string`

When a Dispute is nested within a charge, it will be charge_id rather than charge object

Jinksi committed:
Move useDisputeAccept hook to `data/disputes/hooks`

Jinksi committed:
Improve translators note with `<em>` element description

Jinksi committed:
Add DisputeDetails tests

Jinksi committed:
Add test for clicking challenge button

Jinksi commented on client/data/disputes/hooks.ts:
```diff
@@ -43,6 +43,27 @@ export const useDispute = (
 	return { dispute, isLoading, doAccept };
 };
 
+export const useDisputeAccept = (
```
This hook is a wrapper for the `acceptTransactionDetailsDispute` action and an `isLoading` listener.

Jinksi committed:
Remove extraneous div

Jinksi commented on client/types/disputes.d.ts:
```diff
@@ -51,11 +64,12 @@ export interface Dispute {
 	fileSize?: Record< string, number >;
 	reason: DisputeReason;
-	charge: Charge;
+	charge: Charge | string;
```
When the `Dispute` is nested within a `Charge`, the `charge` property is a charge id (`ch_`) rather than a dispute object.

Jinksi committed:
Add test for dispute with staged evidence

Jinksi committed:
Add dispute action tracks events

Jinksi committed:
Add missing full-stop to accept dispute modal text

Jinksi commented:
**Before review:** I've noticed that the accept dispute modal will show as `undefined` if the dispute is an Inquiry since a fee has not been applied to the transaction yet.

![image](https://github.com/Automattic/woocommerce-payments/assets/3147296/9119b7af-975c-45a7-854a-c07269d70541)

@nikkivias should we modify the text in the accept dispute modal for inquiries? I believe the outcome will be the same.

We don't have the fee amount yet since it has not been applied to the transaction for inquiries. But we can probably source the expected fee amount from somewhere else if we want to show the fee here.

Jinksi committed:
Add currency rendering test for accept dispute modal

Jinksi committed:
Add test for Inquiry accept dispute modal

Jinksi commented:
Marking as `blocked` while we figure out content and/or actions for Inquiries. https://github.com/Automattic/woocommerce-payments/pull/7093#issuecomment-1700505732

Note, this is being discussed in Slack here: p1693496507185109-slack-C03LA109BKM

Jinksi committed:
Merge branch 'develop' into fix/6926-transaction-dispute-details-actions

Jinksi committed:
Merge branch 'develop' into fix/6926-transaction-dispute-details-actions

Jinksi committed:
Merge branch 'develop' into fix/6926-transaction-dispute-details-actions

Jinksi committed:
Fix formatting after merge conflict resolution

Jinksi committed:
Merge branch 'develop' into fix/6926-transaction-dispute-details-actions

Jinksi committed:
Fix duplicate import after merge conflict resolution

haszari commented:
> Before review: I've noticed that the accept dispute modal will show as undefined if the dispute is an Inquiry since a fee has not been applied to the transaction yet.

> We don't have the fee amount yet since it has not been applied to the transaction for inquiries. But we can probably source the expected fee amount from somewhere else if we want to show the fee here.

@Jinksi @nikkivias Could we just gracefully handle a missing fee value in the message logic? Something like the following:

- If it's a regular dispute, and/or we know the fee show: `The disputed amount will be returned to the cardholder, with a $17.23 dispute fee deducted from your account.`
- If it's an inquiry, or we don't know the fee show: `The disputed amount will be returned to the cardholder, with a dispute fee deducted from your account.`

Jinksi committed:
Merge branch 'develop' into fix/6926-transaction-dispute-details-actions

Jinksi commented:
> Could we just gracefully handle a missing fee value in the message logic?


@haszari there may be more problems with this statement for inquiries, e.g. "the disputed amount will be returned to the cardholder" is not true for inquiries.

Also, the action "Accept" may not be available or relevant for inquiries – this may represent a different action (e.g. "Refund").

Slack thread remains in progress, partially about this issue: p1693496507185109-slack-C03LA109BKM.

Jinksi committed:
Merge branch 'develop' into fix/6926-transaction-dispute-details-actions

Jinksi committed:
Merge branch 'develop' into fix/6926-transaction-dispute-details-actions

Jinksi commented:
Note this PR/issue is no longer blocked, as CTA buttons and modals for inquiries will be handled in issue #7193. I'll tidy up this PR and prepare it for another review. cc @nikkivias @haszari

Jinksi committed:
Merge branch 'develop' into fix/6926-transaction-dispute-details-actions

Jinksi committed:
Don't render dispute actions for inquiries

Jinksi committed:
Add more info to acceptTransactionDetailsDispute comment

Jinksi committed:
Rename onSubmit → onAccept to name intent

Jinksi committed:
Add dispute_status prop to tracks events

Jinksi committed:
Update dispute hook comments

Jinksi committed:
Cleanup tests and remove flaky test

Jinksi committed:
Move challenge click handler to func improve code clarity

Jinksi committed:
Remove redundant `null` return

Jinksi committed:
Merge branch 'develop' into fix/6926-transaction-dispute-details-actions

Jinksi committed:
Merge branch 'develop' into fix/6926-transaction-dispute-details-actions

Jinksi committed:
Merge branch 'develop' into fix/6926-transaction-dispute-details-actions

Jinksi committed:
Merge branch 'develop' into fix/6926-transaction-dispute-details-actions

Jinksi committed:
Merge branch 'develop' into fix/6926-transaction-dispute-details-actions

Jinksi committed:
Move DisputeActions logic and markup into parent component

Jinksi commented:
Marking in-progress while I make changes as per feedback in PR review https://github.com/Automattic/woocommerce-payments/pull/7047#pullrequestreview-1625765513

Jinksi committed:
Allow native browser navigation for challenge button

Jinksi committed:
Add test for challenge button click

Jinksi committed:
Move accept click handler inline to improve readability

Jinksi committed:
Use `showDisputeActions` variable to improve readability

Jinksi committed:
Add descriptive comment about challenge link

Jinksi committed:
Rename test `index.tsx` to `index.test.tsx` for consistency

brucealdridge reviewed (changes requested):
Challenge dispute → Dispute challenge screen ✅
Accept dispute → A modal will be shown. ✅
On error, a snackbar will show an error message in the bottom-left corner. ✅
The dispute should be accepted ✅
The Transaction Details screen will be hydrated with new data (Disputed: Lost status chip). ✅
The dispute actions should no longer be available. ❌ 
The dispute actions should not be visible when won, lost or under review. ❌
View a disputed transaction for a dispute where a challenge has been staged (click "Save for Later" on the dispute challenge form.) ✅

When testing I was able to see the buttons after the dispute was lost or won.
![Screenshot 2023-09-19 at 3 59 02 PM](https://github.com/Automattic/woocommerce-payments/assets/57298/2d9a4396-2365-4c60-ba5b-13051b638e3c)

Jinksi committed:
Update styles on mobile screens to match design

Jinksi committed:
Add tests for when actions are not allowed

Jinksi commented:
> The dispute actions should no longer be available. ❌
> The dispute actions should not be visible when won, lost or under review. ❌

This surprises me, since the parent won't render the component at all if the `isAwaitingResponse` return false.
https://github.com/Automattic/woocommerce-payments/blob/d8ca8e6e7b2c9f3a63760c80b24e4e4db436691f/client/payment-details/summary/index.tsx#L383-L387

I'll have a look. Thanks for testing @brucealdridge !

brucealdridge commented on client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:
```diff
@@ -61,6 +88,156 @@ const DisputeAwaitingResponseDetails: React.FC< Props > = ( { dispute } ) => {
+					{ /* Dispute Actions */ }
+					{ showDisputeActions && (
+						<Flex justify="start">
+							<Link
+								href={
+									// Prevent the user navigating to the challenge screen if the accept request is in progress.
+									isLoading
+										? ''
+										: getAdminUrl( {
+												page: 'wc-admin',
+												path:
+													'/payments/disputes/challenge',
+												id: dispute.id,
+										  } )
+								}
+							>
+								<Button
```
Is there a reason for a button inside a link here? Would it be better to just have a button with a `href` attribute on it? 

There is the `wcTracks` event on the `<Button>` and the `href` on the `<Link>`

Can we combine these so there are two actions on the one element, rather than two actions over two elements?

Jinksi replied:
Ah, yeh this pattern came out of the discussion here: https://github.com/Automattic/woocommerce-payments/pull/7047#discussion_r1328174632.

We want the client-side routing AND native browser navigation via href when CMD+clicking. `Link` offers this, whereas `Button` doesn't – but we want the `Button` UI.

I'll consider if there's another way, perhaps there is a component that I don't know about.

Jinksi replied:
I've attempted a compromise in 3709cc0 by using a `Link` with a `span` child for the button styles.

```tsx
<Link
	href={ getAdminUrl( ... ) }
	onClick={ () => { wcpayTracks.recordEvent( ... ); } }
>
	<span className="components-button is-primary">
		I am styled like a Button
```

One risk with this approach is that we're using classNames for styling `components-button` detached from the `Button` component.

I'm curious to get your thoughts on this problem, @haszari, as you noticed this issue in a recent PR: https://github.com/Automattic/woocommerce-payments/pull/7047#discussion_r1328174632.

If I'm thinking about the project goals, getting this button right is critical. Using client-side routing is desirable, IMO, since the merchant will see the challenge form more quickly and smoothly → more challenges being initiated.

haszari replied:
> If I'm thinking about the project goals, getting this button right is critical. Using client-side routing is desirable, IMO, since the merchant will see the challenge form more quickly and smoothly → more challenges being initiated.

💯 totally agree. If we can speed up state transitions, that should improve "conversions" (responses to disputes), and of course make merchants' lives easier and more efficient. 

> One risk with this approach is that we're using classNames for styling components-button detached from the Button component.

I don't think using a raw `<span>` is much leaner than using a `<button>`. I'd lean towards using components as is (not borrowing classes or other hackery), and if there's a gap in the library, follow up by logging issues (and submitting PRs?) so the component library meets our needs. So personally I prefer the original `<Link><Button /></Link>

This reveals a potential UX/design issue – we might be using `Button` inappropriately when we should be using `Link`.

A bigger topic really, might be a good discussion on P2 e.g. codep2 or with other front end experts :)

Jinksi replied:
Thanks for your feedback @haszari and @brucealdridge.

Based on this discussion, I'm going to proceed with the `<Link><Button /></Link>` pattern to get this PR shipped (203f161 & 163514b).

I'll log a follow-up issue and move this conversation elsewhere.

brucealdridge commented on client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:
```diff
@@ -61,6 +88,156 @@ const DisputeAwaitingResponseDetails: React.FC< Props > = ( { dispute } ) => {
+												sprintf(
+													/* translators: %s: dispute fee, <em>: emphasis HTML element. */
+													__(
+														'Accepting the dispute marks it as <em>Lost</em>. The disputed amount will be returned to the cardholder, with a %s dispute fee deducted from your account.',
```
I think we need to review this language. This implies the fee will only be deducted if they accept the dispute. When the fee has already been deducted.

Jinksi replied:
I'll log a follow-up issue to allow this PR to be shipped and we can iterate on the wording.

Jinksi replied:
New issue for this: #7254

Jinksi committed:
Remove `Button` from `Link` to simplify code

but keep functionality and styles

Jinksi committed:
Fix tests after changing Button → Link

brucealdridge commented:
> This surprises me, since the parent won't render the component at all if the `isAwaitingResponse` return false.

Hmmmm, I'm sure I did a pull just before testing. Doing a pull again says I was running be9ff75. False alarm sorry, all working as expected.

brucealdridge reviewed (approved):
Retested after confirming I was checking an old commit :doh!: 
All cases working as expected.

Jinksi committed:
Revert "Fix tests after changing Button → Link"

This reverts commit 340b5a172abea2e9e1bfe8f28bd61b1cb0fccda6.

Jinksi committed:
Revert "Remove `Button` from `Link` to simplify code"

This reverts commit 3709cc0b8bb0899dd0cb011648e7d891a36fba5d.

Jinksi committed:
Merge branch 'develop' into fix/6926-transaction-dispute-details-actions
